crypto/dh/dh_lib.c

```diff
@@ -246,14 +246,21 @@ void DH_get0_pqg(const DH *dh, BIGNUM **p, BIGNUM **q, BIGNUM **g)
 int DH_set0_pqg(DH *dh, BIGNUM *p, BIGNUM *q, BIGNUM *g)
 {
     /* q is optional */
-    if (p == NULL || g == NULL)
+    if ((p == NULL && dh->p == NULL)
+        || (g == NULL && dh->g == NULL))
```

Comment:
The advantage of my proposed approach is that it doesn't change the semantics of API at all. It just detects and reports an error condition that at least one experienced developer ran into that would otherwise (probably) result in a seg fault. As it's not hard to check for, and isn't any more expensive to do the check than what we've already got, why would we not do it? Anyway, I'm fine with leaving it as is if that is the consensus - but in any case I think it wouldn't hurt to try and make the documentation even more explicit.